docs/developers/arch/adr-003.md

```diff
@@ -0,0 +1,114 @@
+# Sidebar application architecture
+
+This document describes a proposed plan to clarify the structure of the Hypothesis client's sidebar application.
+
+## Context
+
+The Hypothesis client consists of three major pieces:
+
+- The "boot" code which loads resources required by other parts of the client
+- The "annotator" code which runs in the context of the host page
+- The "sidebar" application which handles fetching, presenting and editing of
+  annotations
+
+The sidebar application is built on Angular JS [1]. Other than the [components
+ADR](adr-001.md) the architecture has evolved in a slightly ad-hoc fashion.
+
+For example, historically the app suffered from a number of problems related to
+performance [2] and testability [3] which motivated the introduction of a
+central state container. However it is not universally used for all the data
+needed to display the UI, creating confusion about what to use for new features.
+
+In addition to inconsistencies, the structure that exists is difficult to
+discern without extensive reading of the code and new or occassional developers
+have found it difficult to grok. Poor naming of some modules also complicates
+understanding. For example the API client is called `store.js` (a reference to
+_Annotator Store_).
+
+## Proposal
+
+Make explicit the implicit structure which the sidebar app's code has gradually
+been evolving toward over the last 12-18 months and to finish the transition.
+The sidebar app will consist of the following layers:
+
+**Store**
+- The _state container_ [4] for the app, built on Redux.
+- This maintains all the state needed to "draw" the user interface in a normalized (ie. avoiding duplication) representation.
+- Organised into modules which maintain different aspects of the state.
+
+**Services**
+- Classes that handle interaction with external actors such as the host page, the annotation service (via the API, real-time API and OAuth endpoints) and local storage, scheduled/periodic tasks.
+- These may maintain transient state (eg. the active `WebSocket` instance), but nothing required to "draw" the UI
```

Comment:
I feel like this point may bear additional clarification as we go: what kind of state is OK? I'm not sure I grok the notion of "transient state" here...what state in the app _isn't_ transient :) ?

Reply:
I think you're right. We'll need to clarify this a bit as we go. Transient might not be the best term. Some examples might paint a clearer picture:

- Things which you couldn't usefully serialise and thus don't make sense to store in the app's "database", such as a `WebSocket` connection
- State which is purely local to one component and doesn't need to be kept if that component is destroyed and later recreated, for example whether the share popup is open on an annotation card
- Animation/transition states